clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/Applications.java

```diff
@@ -145,46 +145,50 @@ public CloudFoundryServerGroup findById(String guid) {
-                        if (cachedApp != null) {
-                          if (!cachedApp
-                              .getUpdatedTime()
-                              .equals(app.getUpdatedAt().toInstant().toEpochMilli())) {
-                            log.trace(
-                                "App '{}' cached version is out of date on foundation '{}'",
-                                app.getName(),
-                                this.account);
-                            return true;
-                          } else {
-                            return false;
-                          }
-                        } else {
-                          log.trace(
-                              "App '{}' not found in cache for foundation '{}'",
-                              app.getName(),
-                              this.account);
-                          return true;
-                        }
-                      })
-                  .map(this::map)
-                  .forEach(sg -> serverGroupCache.put(sg.getId(), sg)));
-
-      forkJoinPool.submit(
-          () ->
-              // execute health check on instances, set number of available instances and health
-              // status
-              newCloudFoundryAppList
-                  .parallelStream()
-                  .forEach(
-                      a ->
-                          serverGroupCache.put(
-                              a.getGuid(), checkHealthStatus(findById(a.getGuid()), a))));
+      forkJoinPool
+          .submit(
+              () ->
+                  newCloudFoundryAppList
+                      .parallelStream()
+                      .filter(
+                          app -> {
+                            CloudFoundryServerGroup cachedApp = findById(app.getGuid());
+                            if (cachedApp != null) {
+                              if (!cachedApp
+                                  .getUpdatedTime()
+                                  .equals(app.getUpdatedAt().toInstant().toEpochMilli())) {
+                                log.trace(
+                                    "App '{}' cached version is out of date on foundation '{}'",
+                                    app.getName(),
+                                    this.account);
+                                return true;
+                              } else {
+                                return false;
+                              }
+                            } else {
+                              log.trace(
+                                  "App '{}' not found in cache for foundation '{}'",
+                                  app.getName(),
+                                  this.account);
+                              return true;
+                            }
+                          })
+                      .map(this::map)
+                      .forEach(sg -> serverGroupCache.put(sg.getId(), sg)))
+          .get();
```

Comment:
This isn't a blocker since the original code is like this but it'd be _really nice_ if the code within the `filter` was broken out into a different method. This bit of code is hard to reason about.

Reply:
I'm not familiar with `ForkJoinPool`. I assume the absence of the call to `get` in the previous implementation was causing the task to end prematurely and resulting in the stale data?

Reply:
Yeah we'll revisit that part in a later change. I believe it's still going through the health check too many times. The `get()` (could have used `join()`) forces the code to wait for the parallel execution to complete.